### DISCUSSION TITLE
如何使用Go实现一个简单的分布式锁?

### DISCUSSION THREAD

**Discussion by beijixi** [RunwangGuo/short-url-4go#2] [Q&A] [answered]
如何使用Go实现一个简单的分布式锁?

**Comment by RunwangGuo** [comment, marked as answer]
可以使用Redis作为后端存储,实现一个基于Redlock算法的分布式锁。以下是一个简单的实现:
package main

import (
 "context"
 "fmt"
 "log"
 "time"

 "github.com/go-redis/redis/v8"
)

type DistributedLock struct {
 client *redis.Client
 key    string
 expire time.Duration
}

func NewDistributedLock(client *redis.Client, key string, expire time.Duration) *DistributedLock {
 return &DistributedLock{
  client: client,
  key:    fmt.Sprintf("%s:%s", "lock", key),
  expire: expire,
 }
}

func (d *DistributedLock) Acquire(ctx context.Context) error {
 for {
  err := d.client.Set(ctx, d.key, "locked", d.expire).Err()
  if err != nil {
   return err
  }

  if d.client.TTL(ctx, d.key).Val() > 0 {
   return nil
  }

  time.Sleep(time.Millisecond * 100)
 }
}

func (d *DistributedLock) Release(ctx context.Context) error {
 return d.client.Del(ctx, d.key).Err()
}

func main() {
 rdb := redis.NewClient(&redis.Options{
  Addr: "localhost:6379",
 })

 lock := NewDistributedLock(rdb, "example_lock", 5*time.Second)

 ctx := context.Background()

 err := lock.Acquire(ctx)
 if err != nil {
  log.Fatalf("Failed to acquire lock: %v", err)
 }

 defer func() {
  lock.Release(ctx)
 }()

 fmt.Println("Lock acquired. Doing some work...")
 time.Sleep(3 * time.Second)
}
